commandapi-core/src/main/java/dev/jorel/commandapi/AbstractCommandAPICommand.java

```diff
@@ -212,29 +247,67 @@ public void register() {
 			CommandAPI.logWarning("Command /" + meta.commandName + " is being registered after the server had loaded. Undefined behavior ahead!");
 		}
 		try {
-			Argument[] argumentsArr = (Argument[]) (args == null ? new AbstractArgument[0] : args.toArray(AbstractArgument[]::new));
+			Argument[] argumentsArray = (Argument[]) (arguments == null ? new AbstractArgument[0] : arguments.toArray(AbstractArgument[]::new));
 
 			// Check IGreedyArgument constraints
-			for (int i = 0, numGreedyArgs = 0; i < argumentsArr.length; i++) {
-				if (argumentsArr[i] instanceof IGreedyArgument) {
-					if (++numGreedyArgs > 1 || i != argumentsArr.length - 1) {
-						throw new GreedyArgumentException(argumentsArr);
+			for (int i = 0, numGreedyArgs = 0; i < argumentsArray.length; i++) {
+				if (argumentsArray[i] instanceof IGreedyArgument) {
+					if (++numGreedyArgs > 1 || i != argumentsArray.length - 1) {
+						throw new GreedyArgumentException(argumentsArray);
+					}
+				}
+			}
+
+			// Check optional argument constraints
+			// They can only be at the end, no required argument can follow an optional argument
+			int firstOptionalArgumentIndex = -1;
+			for (int i = 0, optionalArgumentIndex = -1; i < argumentsArray.length; i++) {
+				if (argumentsArray[i].isOptional()) {
+					if (firstOptionalArgumentIndex == -1) {
+						firstOptionalArgumentIndex = i;
 					}
+					optionalArgumentIndex = i;
+				} else if (optionalArgumentIndex != -1) {
+					// Argument is not optional
+					throw new OptionalArgumentException();
 				}
 			}
 
 			// Assign the command's permissions to arguments if the arguments don't already
 			// have one
-			for (Argument argument : argumentsArr) {
+			for (Argument argument : argumentsArray) {
 				if (argument.getArgumentPermission() == null) {
 					argument.withPermission(meta.permission);
 				}
 			}
 
+			// Create a List of arrays that hold arguments to register optional arguments
+			// if optional arguments have been found
+			List<Argument[]> argumentsToRegister = new ArrayList<>();
+			if (firstOptionalArgumentIndex != -1) {
+				for (int i = 0; i <= argumentsArray.length; i++) {
+					if (i == firstOptionalArgumentIndex - 1) {
+						Argument[] requiredArguments = (Argument[]) new AbstractArgument[i];
+						System.arraycopy(argumentsArray, 0, requiredArguments, 0, i);
+						argumentsToRegister.add(requiredArguments);
+					} else if (i >= firstOptionalArgumentIndex) {
+						Argument[] optionalArguments = (Argument[]) new AbstractArgument[i];
+						System.arraycopy(argumentsArray, 0, optionalArguments, 0, i);
+						argumentsToRegister.add(optionalArguments);
+					}
+				}
+			}
```

Comment:
CodeFactor complains about this causing this method to be fairly complex. I initially dismissed this, but upon further inspection I think it's right. We have two parts for optional arguments here:
- Setting `firstOptionalArgumentIndex` and throwing `OptionalArgumentException`
- Generating the `argumentsToRegister` list

We could probably put both of these in one (private) function and call that instead of having these two parts.

Reply:
While I think one method might be good, I decided to make this two blocks because between those two block, permissions are linked to arguments.
(I hope this was understandable...)

Reply:
You can set the permissions first, then handle optional arguments after that.

Reply:
Yeah...
Give me a moment. I'll change that!